impl Trait not allowed after updating tracing-attributes:0.1.12

## Bug Report

Hey :wave: thanks for taking the time to look at the following issue :blush: 

### Version

The versions are:
`tracing = 0.1.23`
`tracing-attributes = 0.1.12`
`tracing-core = 0.1.17`

Here's the [`Cargo.lock`](https://github.com/radicle-dev/tracing-bug/blob/master/Cargo.lock) of the reproducible.

### Platform

`Linux haptop 5.10.0-rc6 #1-NixOS SMP Sun Nov 29 23:50:50 UTC 2020 x86_64 GNU/Linux`

### Description

We're using a combination of `tracing::instrument` while also returning a trait object, i.e. `impl`. When this combined with `Result` and using `err` in `tracing::instrument` then we get the following compiler error:
```
error[E0562]: `impl Trait` not allowed outside of function and inherent method return types
  --> src/main.rs:22:30
   |
22 |     fn test(&self) -> Result<impl Iterator<Item = u32>, Error> {
   |                              ^^^^^^^^^^^^^^^^^^^^^^^^^
```

This was working before and it seems that an update of `tracing-attributes` caused the code not to compile. It's probably worth noting that this isn't an issue without the `Result` and `err` pragma.

Here's a minimal example of a project not compiling https://github.com/radicle-dev/tracing-bug. If you switch to the `works` branch, https://github.com/radicle-dev/tracing-bug/tree/works, then you can see that using `tracing-attributes = "=0.1.11"` allows the code to compile again. 

If there's any more detail that would be helpful, please let me know :v: Thanks again!

--- a/tracing-attributes/src/lib.rs
+++ b/tracing-attributes/src/lib.rs
@@ -469,7 +469,7 @@ fn gen_body(
         quote_spanned!(block.span()=>
             let __tracing_attr_span = #span;
             let __tracing_attr_guard = __tracing_attr_span.enter();
-            match move || #return_type #block () {
+            match (move || #block)() {
                 #[allow(clippy::unit_arg)]
                 Ok(x) => Ok(x),
                 Err(e) => {

--- a/tracing-attributes/src/lib.rs
+++ b/tracing-attributes/src/lib.rs
@@ -469,6 +469,7 @@ fn gen_body(
         quote_spanned!(block.span()=>
             let __tracing_attr_span = #span;
             let __tracing_attr_guard = __tracing_attr_span.enter();
+            #[allow(clippy::redundant_closure_call)]
             match (move || #block)() {
                 #[allow(clippy::unit_arg)]
                 Ok(x) => Ok(x),

--- a/tracing-attributes/tests/err.rs
+++ b/tracing-attributes/tests/err.rs
@@ -137,3 +137,34 @@ fn test_mut_async() {
     });
     handle.assert_finished();
 }
+
+#[test]
+fn impl_trait_return_type() {
+    // Reproduces https://github.com/tokio-rs/tracing/issues/1227
+
+    #[instrument(err)]
+    fn returns_impl_trait(x: usize) -> Result<impl Iterator<Item = usize>, String> {
+        Ok(0..x)
+    }
+
+    let span = span::mock().named("returns_impl_trait");
+
+    let (collector, handle) = collector::mock()
+        .new_span(
+            span.clone()
+                .with_field(field::mock("x").with_value(&format_args!("10")).only()),
+        )
+        .enter(span.clone())
+        .exit(span.clone())
+        .drop_span(span)
+        .done()
+        .run_with_handle();
+
+    with_default(collector, || {
+        for _ in returns_impl_trait(10).unwrap() {
+            // nop
+        }
+    });
+
+    handle.assert_finished();
+}

--- a/tracing-attributes/tests/instrument.rs
+++ b/tracing-attributes/tests/instrument.rs
@@ -200,3 +200,32 @@ fn methods() {
 
     handle.assert_finished();
 }
+
+#[test]
+fn impl_trait_return_type() {
+    #[instrument]
+    fn returns_impl_trait(x: usize) -> impl Iterator<Item = usize> {
+        0..x
+    }
+
+    let span = span::mock().named("returns_impl_trait");
+
+    let (collector, handle) = collector::mock()
+        .new_span(
+            span.clone()
+                .with_field(field::mock("x").with_value(&format_args!("10")).only()),
+        )
+        .enter(span.clone())
+        .exit(span.clone())
+        .drop_span(span)
+        .done()
+        .run_with_handle();
+
+    with_default(collector, || {
+        for _ in returns_impl_trait(10) {
+            // nop
+        }
+    });
+
+    handle.assert_finished();
+}

